<!doctype html>
<html>
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

    <title>Nanopolish</title>

    <link rel="stylesheet" href="../reveal.js/css/reveal.css">
    <link rel="stylesheet" href="../reveal.js/css/theme/sky.css">
    <link rel="stylesheet" href="nanopolish.css">

    <!-- Theme used for syntax highlighting of code -->
    <!--<link rel="stylesheet" href="../reveal.js/lib/css/atom-one-dark.css">-->

    <!-- Printing and PDF exports -->
    <script>
        var link = document.createElement( 'link' );
        link.rel = 'stylesheet';
        link.type = 'text/css';
        link.href = window.location.search.match( /print-pdf/gi ) ? '../reveal.js/css/print/pdf.css' : '../reveal.js/css/print/paper.css';
        document.getElementsByTagName( 'head' )[0].appendChild( link );
    </script>
    <script async defer src="https://buttons.github.io/buttons.js"></script>
</head>
<body>
<div class="reveal">
    <div class="slides">

        <!-- =========== SLIDE 1 Title =========== -->
        <section>
            <h2 class="title">Nanopolish</h2>
            <p><a href="https://github.com/jts/nanopolish" target="_blank">https://github.com/jts/nanopolish</a></p>
            <p>
                Author: Jared Simpson
            </p>
            <a class="github-button" href="https://github.com/jts/nanopolish" aria-label="View on GitHub" data-size="large">View on GitHub</a>
            <p>
                <br> <br>
                <small>Slides: Michael Hall</small>
                <br>
                <a class="github-button" href="https://github.com/mbhall88" aria-label="Follow @mbhall88 on GitHub">Follow @mbhall88</a>
                <a href="https://twitter.com/mbhall88?ref_src=twsrc%5Etfw" class="twitter-follow-button" data-show-count="false">Follow @mbhall88</a><script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
            </p>
        </section>

        <!-- =========== SLIDE 2 Key Message =========== -->
        <section>
            <div>
                <p>Software tool for using signal-level nanopore data to:</p>
                <ul>
                    <li><p>Improve a consensus sequence</p></li>
                    <li class="fragment"><p>Call variants</p></li>
                    <li class="fragment"><p>Detect methylated bases</p></li>
                </ul>
            </div>
        </section>

        <!-- =========== SLIDE 3 Program Overview =========== -->
        <section data-markdown>
            <textarea data-template>
                ```
                nanopolish extract: extract reads in FASTA or FASTQ format from
                    a directory of FAST5 files

                nanopolish call-methylation: predict genomic bases that may be
                    methylated

                nanopolish variants: detect SNPs and indels with respect to a
                    reference genome

                nanopolish variants --consensus: calculate an improved consensus
                    sequence for a draft genome assembly

                nanopolish eventalign: align signal-level events to k-mers of a
                    reference genome
                ```
            </textarea>
        </section>

        <!-- =========== SLIDE 4 Modelling events =========== -->
        <section>

            <!-- 4A -->
            <section>
                <h3 class="header">Modelling events</h3>
                <p>
                    The probability of observing an event $e_i$ given that the 5-mer $k$ is the true sequence is
                </p>
                $$P(e_i|k, \Theta_v) = \mathcal{N}(a+b\mu_k+ct_i,(d\theta_k)^2)$$
                <p class="fragment">
                    <br>
                    <small>
                        $e_i=$ measured current level for event $i$, <br>
                        $t_i=$ time at which event $i$ is observed, <br>
                        $k=$ k-mer, <br>
                        $\mu_k=$ mean current level for k-mer $k$, <br>
                        $\theta_k=$ standard deviation for current level of $k$, <br>
                        $a=$ shift parameter, $b=$ scale parameter, $c=$ drift parameter, $d=$ variance parameter. <br>
                        As $a-d$ are not provided for R9 data (RNN basecaller), set $c=0$ and solve a linear least
                        squares regression to calculate $a, b, $ and $d$.
                    </small>
                </p>

            </section>

            <!-- 4B -->
            <section>
                <h3 class="header">Modelling events</h3>
                <p>
                    In order to model a sequence $S$ we represent it as a sequence of 5-mers $S=(s_1, s_2,â€¦,s_m)$. <br>
                    In a perfect world there would be a one-to-one mapping of events $e$ to 5-mers $s$. In this utopia
                    the probability of the data could be calculated as
                </p>
                $$P(e_1,e_2,...,e_n|s_1,s_2,...,s_n,\Theta)=\prod_{i=1}^nP(e_i|s_i,\mu_{s_i},\sigma_{s_i})$$
            </section>

            <!-- 4C -->
            <section>
                <h3 class="header">Modelling events - example</h3>
                <p>
                    In reality though there are two other situations which can arise. <br>
                    <br>
                    <strong>Skip event</strong> - 5-mer has passed through the pore without being detected <br>
                    <br>
                    <strong>Split event</strong> - 5-mer has emitted multiple events.
                </p>
            </section>

            <!-- 4D -->
            <section>
                <h3 class="header">Modelling events - example</h3>
                <p>
                    $S=(ACGTA,CGTAA,GTAAC,TAACT)$ <br> <br>
                    Starting at $ACGTA$ we can either go to the next 5-mer, $CGTAA$, <strong>skip</strong> to $GTAAC$,
                    or <strong>split</strong> the event and stay where we are - emitting another event.
                </p>
                <p>
                    <small>Later in development there has been two other states added. <strong>Softclip</strong> and
                    <strong>bad</strong>.</small>
                </p>
            </section>

            <!-- 4E -->
            <section class="slide4e">
                <h3 class="header">Modelling events - Enter the HMM</h3>
                <p>
                    $\pi$ is a path through the system (hidden state), where an entry is a pair $\pi_i=(i, s_j)$ which says
                    at step $i$ the system emitted event $i$ from 5-mer $s_j$.

                    The transition probability is calculated as $P(\pi_i|\pi_{i-1},S)$. <br>
                </p>

               <div id="eq4e">
                   $$
                    P(e_1,e_2,...,e_n|s_1,s_2,...,s_n,\Theta)=\prod_{i=1}^nP(e_i|s_i,\mu_{s_i},\sigma_{s_i})P(\pi_i|\pi_{i-1}, S)
                    $$
               </div>
            </section>

            <!-- 4F -->
            <section>
                <h3 class="header">Modelling events - Enter the HMM</h3>
                <p>
                    We can then calculate the probability of the data given $S$ by summing over all the possible paths
                    with the Forward algorithm
                </p>
                <div>
                    $$
                    P(e_1,e_2,...,e_n|S,\Theta)=\sum_{\pi}P(\pi, e_1, e_2,...,e_n|S,\Theta)
                    $$
                </div>
            </section>
        </section>

        <!-- =========== SLIDE 5 HMM Profile =========== -->
        <section>

            <!-- 5A -->
            <section>
                <h3 class="header">HMM Profile</h3>
                <div class="hmm_wrap">
                    <div class="hmm">
                        <img src="imgs/hmm.png" alt="HMM">
                    </div>
                    <div class="hmm">
                        <p>
                            $E$ states are split events. <br> <br>
                            $K$ states are skip events. <br> <br>
                            $M$ states are normal movement to next k-mer.
                        </p>
                    </div>
                </div>
            </section>

            <!-- 5B -->
            <section>
                <h3 class="header">HMM Profile - Transition Probabilities</h3>
                <div class="hmm_wrap">
                    <div class="hmm">
                        <img src="imgs/hmm.png" alt="HMM">
                    </div>
                    <div class="hmm">
                        <p id="transition_equations">
                            \[\begin{align*}
                            &t(M_{s_{i-1}}\rightarrow K_{s_i})=f(|\mu_{s_{i-1}}-\mu_{s_i}|) \\

                            &t(M_{s_{i-1}}\rightarrow E_{s_{i-1}})=p_{me}(1-t(M_{s_{i-1}}\rightarrow K_{s_i})) \\

                            &tM_{s_{i-1}}\rightarrow M_{s_i})=1-t(M_{s_{i-1}}\rightarrow K_{s_i})-t(M_{s_{i-1}}\rightarrow E_{s_{i-1}})
                            \\

                            &t(E_{s_{i-1}}\rightarrow E_{s_{i-1}})=p_{ee} \\

                            &t(E_{s_{i-1}}\rightarrow M_{s_i})=1-t(E_{s_{i-1}}\rightarrow E_{s_{i-1}}) \\

                            &t(K_{s_{i-1}}\rightarrow K_{s_i})=f(|\mu_{s_{i-1}}-\mu_{s_i}|) \\

                            &t(K_{s_{i-1}}\rightarrow M_{s_i})=1-t(K_{s_{i-1}}\rightarrow K_{s_i}) \\
                            \end{align*} \]
                        </p>
                    </div>
                </div>
            </section>

            <!-- 5C -->
            <section>
                <h3 class="header">HMM Profile - Transition Probabilities</h3>
                <div class="hmm_wrap">
                    <div class="hmm">
                        <img src="imgs/hmm.png" alt="HMM">
                    </div>
                    <div class="hmm">
                        <p id="emission_equations">
                            \[\begin{align*}
                            &P(e_i|\pi_k=(i,M_{s_j}))=\mathcal{N}(\mu_{s_j},\sigma^2_{s_j}) \\

                            &P(e_i|\pi_k=(i,E_{s_j}))=\mathcal{N}(\mu_{s_j},(v\sigma_{s_j})^2)
                            \end{align*} \]
                        </p>
                    </div>
                </div>
            </section>
        </section>

        <!-- =========== SLIDE 6 Computing Consensus =========== -->
        <section>
            <!-- 6A -->
            <section>
                <h3 class="header">Improved consensus sequence</h3>
                <p>
                    <img src="imgs/nanopolish_analysis_workflow.png" alt="" id="analysis_workflow">
                </p>
            </section>

            <!-- 6B -->
            <section>
                <h3 class="header">Improved consensus sequence</h3>
                <div class="lists">
                    <ol>
                        <li>
                            Uncorrected reads mapped to draft genome $G$
                        </li>
                        <li class="fragment">
                            $G$ is divided into 10Kbp segments with 200bp overlaps.
                        </li>
                        <li>
                            Anchors $A$ selected every 50bp within each segment.
                        </li>
                        <li class="fragment">
                            Compute new sequence between $A_i$ and $A_{i+2}$.
                            Progressively improves event-to-assembly alignment.
                        </li>
                    </ol>
                </div>
            </section>

            <!-- 6C -->
            <section>
                <h3 class="header">Improved consensus sequence</h3>
                <p>
                    Let $\mathcal{D}=\{e_{i},e_{i+1},â€¦,e_{j}\}$ be the set of event subsequences anchored between $A_i$ and
                    $A_{i+2}$. <br>
                    The algorithm tries to compute a consensus sequence $C$ that maximises the probability of data
                    $\mathcal{D}$. <br>
                </p>
            </section>

            <!-- 6D -->
            <section>
                <h3 class="header">Improved consensus sequence</h3>
                <p>
                    This is done by taking the consensus sequence $\mathcal{C}$ between the anchors and proposing a set
                    of candidate sequences. <br>
                    Use the HMM to select the sequence that maximises the probability of the data. <br>
                    As calculating all possible permutations of $\mathcal{C}$ would be too large, <strong>two methods are used
                    to calculate candidates sequences</strong>.
                </p>
            </section>

            <!-- 6E -->
            <section>
                <h3 class="header">Improved consensus sequence</h3>
                <p>
                     <strong>Block replacement</strong> makes edits based on sampling alternative substrings from the
                    reads. <br>

                    <strong>Mutation</strong> (inspired by Quiver) calculates all substrings within one edit distance
                    of the consensus subsequence <br> <br> <br>
                    <small>See
                    <a href="https://media.nature.com/original/nature-assets/nmeth/journal/v12/n8/extref/nmeth.3444-S1.pdf">Supplementary Methods</a>
                    for a more in-depth description of these two methods.</small>

                </p>
            </section>

            <!-- 6F -->
            <section>
                <h3 class="header">Improved consensus sequence</h3>
                <div class="lists">
                    <ol>
                        <li>Calculate all <code>block replacement</code> candidate sequences.</li>
                        <li class="fragment">Calculate the probability of all sequences and select the best candidate $C'$.</li>
                        <li class="fragment">If $C' \neq C$ then $C\leftarrow C'$ and the process is continued.</li>
                    </ol>
                </div>
            </section>

            <!-- 6G -->
            <section>
                <h3 class="header">Improved consensus sequence</h3>
                <div class="lists">
                    <ol>
                        <li value="4">When no improvements are made the process is repeated, but with <code>mutation</code>.</li>
                        <li value="5" class="fragment">Once <code>mutation </code>finishes, the consensus sequence is updated.</li>
                    </ol>
                    <p>This process is run on all 10Kbp segments. They are then joined together at the end using the 200bp overlaps.</p>
                </div>
            </section>
        </section>

        <!-- =========== SLIDE 7 Variant Calling =========== -->
        <section>
            <!-- 7A -->
            <section>
                <h3 class="header">Variant calling</h3>
                <div class="lists">
                    <ol>
                        <li>
                            Find candidate variants from alignment. Positions with >20x coverage and minor allele
                            frequency at least 0.2
                        </li>
                        <li class="fragment">Batch into clusters separated by at least 10bp.</li>
                        <li class="fragment">Generate all haplotypes for cluster and compute likelihood for each.</li>
                        <li class="fragment">Highest likelihood is output to VCF (if not reference).</li>
                    </ol>
                </div>
            </section>

            <!-- 7B -->
            <section>
                <h3 class="header">Variant calling - haplotype generation</h3>
                <div>
                    <p>Candidates had to be clustered due to context of bases in the pore.</p>
                    <p class="fragment">
                        For candidate haplotype generation, all combinations of a cluster are generated, $2^n$, where $n$ is
                        the number of candidate variants.
                    </p>
                    <p class="fragment">Likelihood of each haplotype is calculated using the HMM.</p>
                </div>

            </section>

            <!-- 7C -->
            <section>
                <h3 class="header">Variant calling</h3>
                <div id="variant_img">
                    <img src="https://media.nature.com/lw926/nature-assets/nature/journal/v530/n7589/images_supplementary/nature16996-sf5.jpg"
                         alt="https://media.nature.com/lw926/nature-assets/nature/journal/v530/n7589/images_supplementary/nature16996-sf5.jpg">
                </div>

            </section>
        </section>

        <!-- =========== SLIDE 8  Methylation detection=========== -->
        <section>

            <!-- 8A -->
            <section>
                <h3 class="header">Methylation detection</h3>
                <p>
                    Training data used was PCR-amplified <i>E. coli</i> DNA that was either treated or untreated with the
                    <code>CpG</code> methyltransferase <i>M.SssI</i>. <br>
                    Found the signal shift was dependent on the position of the 5-mC in a given k-mer.
                </p>
                <p class="fragment">
                    Methylation calling algorithm follows similar steps to variant calling. In this case the candidates
                    are clusters based on presence of <code>CG</code> dinucleotides.
                </p>
            </section>

            <!-- 8B -->
            <section>
                <h3 class="header">Methylation detection</h3>
                <div>
                    <img id="meth_pos" src="imgs/methylation_position.jpg" alt="Methylation k-mer position">
                </div>
            </section>

            <!-- 8C -->
            <section>
                <h3 class="header">Methylation detection</h3>
                <p>
                    One limitation: can't distinguish mixed methylation in close
                    proximity. Assigns the same methylation status to <code>CpG</code> sites within 10bp of each other.
                </p>
                <p class="fragment">
                    In paper the method achieved 94% accuracy and called 77% of sites.
                </p>
                <p class="fragment">
                    Natural DNA can have DNA damage and 5-hydroxy-mC which this model does not account for
                    - could reduce accuracy.
                </p>
            </section>

            <!-- 8D -->
            <section>
                <h3 class="header">Methylation detection</h3>
                <p>
                    The log-likelihood ratio assessment compares methylated and unmethylated versions of the k-mer as
                </p>
                $$
                \sum^2_{j=1}\log(\mathcal{L}(S_M|e_{j,1},...,e_{j,n_j}))- \log(\mathcal{L}(S_R|e_{j,1},...,e_{j,n_j}))
                $$
                <p>
                    <small>
                        $S_R=$ substring of reference genome containing at least one <code>CG</code> dinucleotide
                        $S_M=$ modified copy of $S_R$ with all <code>CG</code> change to <code>MG</code>.
                    </small>
                </p>
            </section>

            <!-- 8E -->
            <section>
                <h3 class="header">Methylation detection</h3>
                <div>
                    <img id="meth_hap" src="imgs/methylation_haplotype.png" alt="Methylation haplotype IGV">
                </div>
            </section>
        </section>

        <!-- =========== SLIDE 9 Recent Updates =========== -->
        <section>
            <h3 class="header">Recent updates</h3>
            <div class="lists">
                <ul>
                    <li>Multi-threaded input gives up to 6 times speed-up.</li>
                    <li>Indexing from sequencing summary file.</li>
                    <li>Dynamic scheduling policy for BAM files increases multi-threading efficiency.</li>
                    <li>Phasing coming soon.</li>
                </ul>
            </div>
        </section>

        <!-- =========== SLIDE 10 References =========== -->
        <section>
            <h3 class="header">References</h3>
            <div class="lists references">
                <ul>
                    <li>
                        Consensus algorithm and first methods: <br>
                        <a href="https://www.nature.com/articles/nmeth.3444" target="_blank">
                            <small>Loman, N. J., Quick, J., & Simpson, J. T. (2015). A complete bacterial genome assembled de novo using only nanopore sequencing data. Nature Methods 64(2), p. 10-12.</small>
                        </a>
                    </li>
                    <li>
                        Variant calling method introduced: <br>
                        <a href="https://www.nature.com/articles/nature16996" target="_blank">
                            <small>Quick, J., Loman, N. J., Duraffour, S., Simpson, J. T., Severi, E., Cowley, L., ... Carroll, M. W. (2016). Real. Nature 64(2), p. 10-12.</small>
                        </a>
                    </li>
                    <li>
                        Methylation detection method introduced: <br>
                        <a href="https://www.nature.com/articles/nmeth.4184" target="_blank"><small>
                            Simpson, J. T., Workman, R. E., Zuzarte, P. C., David, M., Dursi, L. J., & Timp, W. (2017). Detecting DNA cytosine methylation using nanopore sequencing. Nature Methods 64(2), p. 10-12.
                        </small></a>
                    </li>
                    <li>
                        The project docs: <br>
                        <a href="https://nanopolish.readthedocs.io/en/latest/index.html" target="_blank">https://nanopolish.readthedocs.io/en/latest/index.html</a>
                    </li>
                    <li>
                        Jared's fantastic lab blog: <br>
                        <a href="https://simpsonlab.github.io/blog/" target="_blank">https://simpsonlab.github.io/blog/</a>
                    </li>
                    <li>
                        The GitHub repository: <br>
                        <a href="https://github.com/jts/nanopolish" target="_blank">https://github.com/jts/nanopolish</a>
                    </li>
                </ul>
            </div>
        </section>

        <!-- =========== SLIDE 11 End =========== -->
        <section>
            <a href="https://mbhall88.github.io/reveal_presentations/nanopolish/nanopolish.html" target="_blank">
                <small>https://mbhall88.github.io/reveal_presentations/nanopolish/nanopolish.html</small>
            </a>
            <a class="github-button" href="https://github.com/mbhall88" data-size="large" aria-label="Follow @mbhall88 on GitHub">Follow @mbhall88</a>
            <a href="https://twitter.com/mbhall88?ref_src=twsrc%5Etfw" class="twitter-follow-button" data-show-count="false" data-size="large">Follow @mbhall88</a><script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
        </section>
    </div>
</div>

<script src="../reveal.js/lib/js/head.min.js"></script>
<script src="../reveal.js/js/reveal.js"></script>

<script>
    // More info about config & dependencies:
    // - https://github.com/hakimel/reveal.js#configuration
    // - https://github.com/hakimel/reveal.js#dependencies
    Reveal.initialize({
        dependencies: [
            { src: '../reveal.js/plugin/markdown/marked.js' },
            { src: '../reveal.js/plugin/markdown/markdown.js' },
            { src: '../reveal.js/plugin/notes/notes.js', async: true },
            { src: '../reveal.js/plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } },
            { src: '../reveal.js/plugin/math/math.js', async: true }
        ],
        transitionSpeed: "fast",
        transition: "none"
    });
</script>

</body>
</html>
