<!doctype html>
<html>
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

    <title>Nanopolish</title>

    <link rel="stylesheet" href="../reveal.js/css/reveal.css">
    <link rel="stylesheet" href="../reveal.js/css/theme/sky.css">
    <link rel="stylesheet" href="nanopolish.css">

    <!-- Theme used for syntax highlighting of code -->
    <!--<link rel="stylesheet" href="../reveal.js/lib/css/atom-one-dark.css">-->

    <!-- Printing and PDF exports -->
    <script>
        var link = document.createElement( 'link' );
        link.rel = 'stylesheet';
        link.type = 'text/css';
        link.href = window.location.search.match( /print-pdf/gi ) ? '../reveal.js/css/print/pdf.css' : '../reveal.js/css/print/paper.css';
        document.getElementsByTagName( 'head' )[0].appendChild( link );
    </script>
</head>
<body>
<div class="reveal">
    <div class="slides">

        <!-- =========== SLIDE 1 Title =========== -->
        <section>
            <h2 class="title">Nanopolish</h2>
            <p><a href="https://github.com/jts/nanopolish" target="_blank">https://github.com/jts/nanopolish</a></p>
            <p>
                <small>Author: Jared Simpson</small>
            </p>
        </section>

        <!-- =========== SLIDE 2 Key Message =========== -->
        <section>
            <div>
                <p>Software tool for using signal-level nanopore data to:</p>
                <ul>
                    <li><p>Improve a consensus sequence</p></li>
                    <li><p>Call variants</p></li>
                    <li><p>Detect methylated bases</p></li>
                </ul>
            </div>
        </section>

        <!-- =========== SLIDE 3 Program Overview =========== -->
        <section data-markdown>
            <textarea data-template>
                ```
                nanopolish extract: extract reads in FASTA or FASTQ format from
                    a directory of FAST5 files

                nanopolish call-methylation: predict genomic bases that may be
                    methylated

                nanopolish variants: detect SNPs and indels with respect to a
                    reference genome

                nanopolish variants --consensus: calculate an improved consensus
                    sequence for a draft genome assembly

                nanopolish eventalign: align signal-level events to k-mers of a
                    reference genome
                ```
            </textarea>
        </section>

        <!-- =========== SLIDE 4 Modelling events =========== -->
        <section>

            <!-- 4A -->
            <section>
                <h3 class="header">Modelling events</h3>
                <p>
                    The probability of observing an event $e_i$ given that the 5-mer $k$ is the true sequence is
                </p>
                    $$P(e_i|k, \Theta)=P(e_i|k, \mu_k, \sigma_k)=\mathcal{N}(\mu_k, \sigma_k^2)$$

            </section>

            <!-- 4B -->
            <section>
                <h3 class="header">Modelling events</h3>
                <p>
                    In order to model a sequence $S$ we represent it as a sequence of 5-mers $S=(s_1, s_2,â€¦,s_m)$. <br>
                    In a perfect world there would be a one-to-one mapping of events $e$ to 5-mers $s$. In this utopia
                    the probability of the data could be calculated as
                </p>
                $$P(e_1,e_2,...,e_n|s_1,s_2,...,s_n,\Theta)=\prod_{i=1}^nP(e_i|s_i,\mu_{s_i},\sigma_{s_i})$$
            </section>

            <!-- 4C -->
            <section>
                <h3 class="header">Modelling events - example</h3>
                <p>
                    In reality though there are two other situations which can arise. <br>
                    <br>
                    <strong>Skip event</strong> - 5-mer has passed through the pore without being detected <br>
                    <br>
                    <strong>Split event</strong> - 5-mer has emitted multiple events.
                </p>
            </section>

            <!-- 4D -->
            <section>
                <h3 class="header">Modelling events - example</h3>
                <p>
                    $S=(ACGTA,CGTAA,GTAAC,TAACT)$ <br> <br>
                    Starting at $ACGTA$ we can either go to the next 5-mer, $CGTAA$, <strong>skip</strong> to $GTAAC$,
                    or <strong>split</strong> the event and stay where we are - emitting another event.
                </p>
            </section>

            <!-- 4E -->
            <section class="slide4e">
                <h3 class="header">Modelling events - Enter the HMM</h3>
                <p>
                    $\pi$ is a path through the system (hidden state), where an entry is a pair $\pi_i=(i, s_j)$ which says
                    at step $i$ the system emitted event $i$ from 5-mer $s_j$.

                    The transition probability is calculated as $P(\pi_i|\pi_{i-1},S)$. <br>
                </p>

               <div id="eq4e">
                   $$
                    P(e_1,e_2,...,e_n|s_1,s_2,...,s_n,\Theta)=\prod_{i=1}^nP(e_i|s_i,\mu_{s_i},\sigma_{s_i})P(\pi_i|\pi_{i-1}, S)
                    $$
               </div>
            </section>

            <!-- 4F -->
            <section>
                <h3 class="header">Modelling events - Enter the HMM</h3>
                <p>
                    We can then calculate the probability of the data given $S$ by summing over all the possible paths
                    with the Forward algorithm
                </p>
                <div>
                    $$
                    P(e_1,e_2,...,e_n|S,\Theta)=\sum_{\pi}P(\pi, e_1, e_2,...,e_n|S,\Theta)
                    $$
                </div>
            </section>
        </section>

        <!-- =========== SLIDE 5 HMM Profile =========== -->
        <section>

            <!-- 5A -->
            <section>
                <h3 class="header">HMM Profile</h3>
                <div id="hmm_wrap">
                    <div class="hmm">
                        <img src="imgs/hmm.png" alt="HMM">
                    </div>
                    <div class="hmm">
                        <p>
                            $E$ states are split events. <br> <br>
                            $K$ states are skip events. <br> <br>
                            $M$ states are normal movement to next k-mer.
                        </p>
                    </div>
                </div>
            </section>

            <!-- 5B -->
            <section>
                <h3 class="header">HMM Profile - Transition Probabilities</h3>
                <div id="hmm_wrap">
                    <div class="hmm">
                        <img src="imgs/hmm.png" alt="HMM">
                    </div>
                    <div class="hmm">
                        <p id="transition_equations">
                            \[\begin{align*}
                            &t(M_{s_{i-1}}\rightarrow K_{s_i})=f(|\mu_{s_{i-1}}-\mu_{s_i}|) \\

                            &t(M_{s_{i-1}}\rightarrow E_{s_{i-1}})=p_{me}(1-t(M_{s_{i-1}}\rightarrow K_{s_i})) \\

                            &tM_{s_{i-1}}\rightarrow M_{s_i})=1-t(M_{s_{i-1}}\rightarrow K_{s_i})-t(M_{s_{i-1}}\rightarrow E_{s_{i-1}})
                            \\

                            &t(E_{s_{i-1}}\rightarrow E_{s_{i-1}})=p_{ee} \\

                            &t(E_{s_{i-1}}\rightarrow M_{s_i})=1-t(E_{s_{i-1}}\rightarrow E_{s_{i-1}}) \\

                            &t(K_{s_{i-1}}\rightarrow K_{s_i})=f(|\mu_{s_{i-1}}-\mu_{s_i}|) \\

                            &t(K_{s_{i-1}}\rightarrow M_{s_i})=1-t(K_{s_{i-1}}\rightarrow K_{s_i}) \\
                            \end{align*} \]
                        </p>
                    </div>
                </div>
            </section>

            <!-- 5C -->
            <section>
                <h3 class="header">HMM Profile - Transition Probabilities</h3>
                <div id="hmm_wrap">
                    <div class="hmm">
                        <img src="imgs/hmm.png" alt="HMM">
                    </div>
                    <div class="hmm">
                        <p id="emission_equations">
                            \[\begin{align*}
                            &P(e_i|\pi_k=(i,M_{s_j}))=\mathcal{N}(\mu_{s_j},\sigma^2_{s_j}) \\

                            &P(e_i|\pi_k=(i,E_{s_j}))=\mathcal{N}(\mu_{s_j},(v\sigma_{s_j})^2)
                            \end{align*} \]
                        </p>
                    </div>
                </div>
            </section>
        </section>

        <!-- =========== SLIDE 6 Computing Consensus =========== -->
        <section>
            <!-- 6A -->
            <section>
                <h4 class="header">Computing improved consensus sequence</h4>
                <p>
                    <img src="imgs/nanopolish_analysis_workflow.png" alt="" id="analysis_workflow">
                </p>
            </section>

            <!-- 6B -->
            <section>
                <h4 class="header">Section 2.2 - Finding all matches within $X$ longer than a minimum length $L$</h4>
                <p>
                    This is implemented in <strong>Algorithm 3</strong>
                    and runs in $O[max(NM, matches)]$ <br> <br>
                    Can be done in the same sweep to calculate $a$ and $d$
                </p>
            </section>
        </section>

        <!-- =========== SLIDE 6 Section 2.3 =========== -->
        <section>

            <!-- 6A -->
            <section>
                <h4 class="header">Section 2.3 - Finding all set-maximal matches within $X$ in linear time</h4>
                <p>
                    Given previous methods, for any sequence $y_i$ at
                    position $k$, it's set-maximal match <strong>must</strong>
                    be immediately prior or following.
                </p>
            </section>

            <!-- 6B -->
            <section>
                <h4 class="header">Section 2.3 - Finding all set-maximal matches within $X$ in linear time</h4>
                <p>
                    Find your interval of longest matches at $k$ - Algorithm 3
                </p>
                <p class="fragment">
                    If $y_j[k] \neq y_i[k]$ then this is set-maximal.
                </p>
                <p class="fragment">
                    Otherwise, the match can be extended and is therefore
                    not set-maximal.
                </p>
                <p class="fragment">
                    Runs in $O(NM)$ - Algorithm 4
                </p>
            </section>
        </section>

        <!-- =========== SLIDE 7 Section 2.4 =========== -->
        <section>
            <h4 class="header">Section 2.4 - Finding all set-maximal matches from a new sequence $z$ to $X$</h4>
            <p>
                Implemented in <strong>Algorithm 5</strong> is a way
                to determine the set-maximal matches of a new sequence(s).
            </p>
            <p class="fragment">
                This can be done in $O(N)$
            </p>
            <p class="fragment">
                I can't quite follow how the algorithm works.
            </p>
        </section>

        <!-- =========== SLIDE 8  Section 2.5 =========== -->
        <section>

            <!-- 8A -->
            <section>
                <h4 class="header">Section 2.5 - Compact representation of $X$</h4>
                <p>
                    Storing $a_k$ and $d_k$ for all $k$ would use more
                    space than the original binary data.
                </p>
                <p class="fragment">
                    Not needed all the time, but are needed if wanting
                    to analyse a new sequence.
                </p>
            </section>

            <!-- 8B -->
            <section>
                <h4 class="header">Section 2.5 - Compact representation of $X$</h4>
                <p>
                    Because there is long stretches of identical values
                    in the sequences the PBWT should be storable in
                    smaller-than-original space.
                </p>
                <p class="fragment">
                    Could be even more space-efficient by storing only
                    a subset of the $a$ and $d$ arrays, say every 32 or
                    64 positions.
                </p>
                <p class="fragment">
                    Makes mention of scope for improving storage of $d$
                    using Huffman coding.
                </p>
            </section>
        </section>

        <!-- =========== SLIDE 9 Results =========== -->
        <section>

            <!-- 9A -->
            <section>
                <h3 class="header">Results</h3>
                <p>
                    <img src="imgs/table1.png" alt="" id="table">
                    Non-linear benefit for compression as $M$ increases.
                </p>
            </section>

            <!-- 9B -->
            <section>
                <h3 class="header">Results</h3>
                <p>
                    <img src="imgs/table2.png" alt="" id="table">
                    Clearly demonstrates the linear time to find
                    set-maximal matches.
                </p>
            </section>

            <!-- 9C -->
            <section>
                <h3 class="header">Results</h3>
                <p>
                    <img src="imgs/table3.png" alt="" id="table">
                    <em>Indexed</em> takes constant time up until the point where
                    it runs out of memory.
                <div id="highlight"></div>
                </p>
            </section>

            <!-- 9D -->
            <section class="slide9d">
                <h3 class="header">Results</h3>
                <p>
                <blockquote>
                    &ldquo;I therefore conclude that the PBWT-based
                    approach can be hunreds of times faster than a
                    direct search approach and find matches in time
                    independent of the reference panel size, as
                    conjectured above, so long as the associated
                    index arrays fit in memory.&rdquo;
                </blockquote>
                </p>
            </section>
        </section>

        <!-- =========== SLIDE 10 Points to ponder =========== -->
        <section>

            <!-- 10A -->
            <section>
                <h3 class="header">Points to Ponder</h3>
                <p>
                <blockquote >
                    &ldquo;Although the algorithms are presented for binary
                    data, they can be extended to multi-allelic
                    data with a little care.&rdquo;
                </blockquote>
                </p>
            </section>

            <!-- 10B -->
            <section>
                <h3 class="header">Points to Ponder</h3>
                <p>
                    All algorithms require exact matching.
                </p>
                <p class="fragment">
                    Could use as a seeding method for statistical
                    genotype algorithms. <br> <br>
                    <small>Mentions possible speedup to BEAGLE</small>
                </p>
            </section>

            <!-- 10C -->
            <section>
                <h3 class="header">Points to Ponder</h3>
                <p>
                    More direct approach would be to work from both ends
                    of the sequences.
                </p>
                <p class="fragment">
                    At some point $k$ we have information about matches in
                    both directions.
                </p>
                <p class="fragment">
                    Can then assign alleles for all sequences at $k$ before
                    moving to next $k$
                </p>
            </section>
        </section>

        <!-- =========== SLIDE 11 Conclusion =========== -->
        <section>
            <section>
                <h2 class="header">Conclusion</h2>
                <p>
                    Series of algorithms to generate prefix array data structures
                    for haplotype sequences and for strong compression.
                </p>
                <p class="fragment">
                    Removes factor of $M$ for search time of new sequences.
                    Allowing matches for ten's of thousands of sequences in
                    minutes.
                </p>
            </section>
            <section>
                <h2 class="header">Conclusion</h2>
                <p>
                    <strong>Importantly</strong>, this could provide a
                    way of storing linkage information for a genome graph. <br> <br>
                    <small>As this can be challenging due to memory limitations.</small>
                </p>
            </section>
        </section>
    </div>
</div>

<script src="../reveal.js/lib/js/head.min.js"></script>
<script src="../reveal.js/js/reveal.js"></script>

<script>
    // More info about config & dependencies:
    // - https://github.com/hakimel/reveal.js#configuration
    // - https://github.com/hakimel/reveal.js#dependencies
    Reveal.initialize({
        dependencies: [
            { src: '../reveal.js/plugin/markdown/marked.js' },
            { src: '../reveal.js/plugin/markdown/markdown.js' },
            { src: '../reveal.js/plugin/notes/notes.js', async: true },
            { src: '../reveal.js/plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } },
            { src: '../reveal.js/plugin/math/math.js', async: true }
        ],
        transitionSpeed: "fast",
        transition: "none"
    });
</script>
</body>
</html>
